### ISSUE TITLE
Idea: use namespace packages in conjunction with plugins

### ISSUE THREAD

**Issue by kne42** [Napari/napari-core#26]
# Introduction
As discussed in our last meeting, [namespace packages](https://packaging.python.org/guides/packaging-namespace-packages/) could be a handy way to ship useful core modules separately for use as standalone packages. This could be extended to all plugins that could be useful outside of the broader Napari ecosystem, such as [`napari-gui`](https://github.com/Napari/napari-gui).

This would provide consistency in `import` statements regardless of how a plugin is installed, as well as allow `pip`-installable plugins to once again reference themselves in absolute imports instead of requiring all imports to be relative.

## What are namespace packages?
Namespace packages are best explained by a simple use-case. Imagine two packages `foo.a` and `foo.b`. `foo.a` and `foo.b` are independent and can be shipped separately. In this case, `foo` would be considered a namespace package. To achieve this, we have the following structure:

```
pkg-foo-a/
    foo/
        a/
            __init__.py
            ...
pkg-foo-b/
    foo/
        b/
            __init__.py
            ...
```

These are known as implicit namespace packages and they were introduced in [PEP 420](https://www.python.org/dev/peps/pep-0420/) (available in Python 3.3+).

In Python <3.3, namespace packages were created using the `pkgutil.extend_path` function under each namespace's `__init__.py`:
```python
# foo/__init__.py
__path__ = __import__('pkgutil').extend_path(__path__, __name__)
```

# Proposal
Implementation of this structure would require two main changes:
- all plugins will fall under the top-level `napari` namespace (as @freeman-lab suggested)
- the [plugin import inception setup](https://github.com/Napari/napari-core/blob/b7e07813322e68cc49a5c5e4cf53dee599fbb792/napari/plugins/__init__.py) will be moved to `napari/__init__/py`

## Refactoring existing plugins
Plugins will no longer be importable under `napari.plugins.<name>` but under `napari.<name>` itself. If installed from `pip`, it will also be importable under `napari.<name>`.

For example, the `napari-gui` repository structure will look like so:
```
napari-gui/
    napari.yml
    napari/
        gui/
            __init__.py
            ...
    ...
```

Its specification will change to:
```
# napari.yml
plugin_info:
  plugin_name: gui
  path: napari/gui/__init__.py
  ...
...
```

 Calls to `setuptools.setup` should specify the `namespace_packages` argument to be `['napari']`.

Plugins already configured with a `setup.py` may not use `setuptools.find_packages()`.

Plugins that do not need to be served as independent `pip` packages should not place their code under a `napari/<name>` directory.

## Implementation
Only a few functions need to be changed, such as `napari/plugins/machinery/_load:get_plugin_namespace`.

The contents of `napari/plugins/__init__.py` will be moved to `napari/__init__.py` **for use with the `plugins` package only**. It will be a `pkgutil`-style namespace package, but will only work if **no other namespace package specifies `napari/__init__.py`**:

```python
# napari/__init__.py
import sys
import pkgutil


class NapariPluginImporter:
    """MetaPathFinder for Napari plugin imports."""
    @classmethod
    def find_spec(cls, fullname, path=None, target=None):
        ...


sys.meta_path.append(NapariPluginImporter)
pkgutil.extend_path(__path__, __name__)
del sys, pkgutil
```

Because packages are queried through the meta-path before the path, discovered packages will always choose their `git`-installed version over the `pip`-installed version, regardless of environment (unless a configuration environment variable is changed).

According to the [PEP 420 specification](https://www.python.org/dev/peps/pep-0420/#specification):
> While looking for a module or package named "foo", for each
> directory in the parent path:
> 
> * If ``<directory>/foo/__init__.py`` is found, a regular package is
>   imported and returned.
> 
> * If not, but ``<directory>/foo.{py,pyc,so,pyd}`` is found, a module
>   is imported and returned.  The exact list of extension varies by
>   platform and whether the -O flag is specified.  The list here is
>   representative.
> 
> * If not, but ``<directory>/foo`` is found and is a directory, it is
>   recorded and the scan continues with the next directory in the
>   parent path.
> 
> * Otherwise the scan continues with the next directory in the parent
>   path.

This means that if the aforementioned plugin-specific `napari/__init__.py` is the only `napari/__init__.py`, it is guaranteed to be executed before all other namespace packages (and therefore always executed whenever an import path begins with `napari`).

# Limitations
- no other namespace package may define `napari/__init__.py`
- due to this implementation, [dynamic path computation](https://www.python.org/dev/peps/pep-0420/#id3) will [no longer be available](https://www.python.org/dev/peps/pep-0420/#migrating-from-legacy-namespace-packages)

**Comment by jni** [comment]
OMG if we'd had namespaces in the early days of scikit-image/learn! :joy: I haven't done a careful review here, but on first pass I love it!

One thing I'd like to consider is to not call the current viewer `napari-gui` but rather `napari-viewer`, reserving `gui` for the full UI with plugins and processing. As we've discussed, just the viewer by itself is super-useful so I'd like that to be a bare-bones standalone thing.

**Comment by kne42** [comment]
@jni that’s a good point but we’ll need to hammer out the details on what constitutes the “viewer” vs the “gui.” For example do we want to include the who layer list thing, drag selection, etc.

**Comment by sofroniewn** [comment]
Have read this over again, and I think it makes a lot of sense. It seems like it will be much easier to integrate packages that are being versioned and distributed separately, which is what we want.

**Comment by jni** [comment]
@stefanv do you have any opinions about namespace packages? I figure scikit-image might have looked at this at some point in the past...

**Comment by stefanv** [comment]
We did have namespaces back in the early skimage days.  In fact, `scikit-image` and `scikit-learn` are the newer names of `scikits.image` and `scikits.learn`.  However, the installation machinery in Python (`easy_install` back then), did not deal with these well, and they often caused problems.  So many problems that we took the trouble of renaming the projects.  I don't know what the latest status is, or whether things have much improved since.

**Comment by jni** [comment]
@stefanv I remember the rename but (a) I have no idea what problems provoked it, and (b) I had always figured that you did something custom and hacky, rather than use an explicit PEP-provided specification. Perhaps I should have known better. But then again... =D

**Comment by stefanv** [comment]
I don't think we did anything strange with namespace packages; I imagine there were problems with multiple versions installed (in-place for scikit-image but normal install of scikit-learn, for example), or similar, while `easy_install` wasn't fully equipped to handle it.  Maybe @cournape, @ogrisel, or @gaelvaroquaux has a better memory?

**Comment by GaelVaroquaux** [comment]
Back in the days, one of the reasons not to use namespace packages is
that they had to override Python's import mechanism which lead to quirks,
and slow down:
http://gael-varoquaux.info/programming/long-syspath-and-consequences-one-more-reason-not-to-use-easy_install.html

I still think that it is something horrible: it's basically an exception
in the way that imports work in Python. I suspect that is will continue
breaking at odd places, because people forget that such mechanism
is possible.

**Comment by AhmetCanSolak** [comment]
Also i want to remind the possible long `sys.path` issues on Windows. I am not sure if using namespace could lead those problems in future but just wanted to point.

**Comment by cournape** [comment]
I think the "new" namespace feature in python 3.3 and above alleviates most of the issues caused in the old days:

1. It does not rely on pkg_resources, and hence does not pay the price of slowing down imports, etc.
2. As it is integrated in python import machinery, I don't think you are paying much by using them (the machinery to look for them is in place whether you are using them or not).

I am not completely sure for 2., as I have not looked at the details of the PEP420 and its implication. But it does add entries on `sys.path` at least.
